app/models/criterion.rb

```diff
@@ -159,4 +162,47 @@ def self.update_assigned_groups_counts(assignment, criterion_ids_by_type = nil)
     #{"AND id IN (#{checkbox_criterion_ids_str})" unless checkbox_criterion_ids_str.empty?}
     UPDATE_SQL
   end
+
+  def replace_marks
+    mark_objects = []
+    # results with specific assignment
+    results = Result
+                .joins(submission: :grouping)
+                .where(groupings: {assignment_id: self.assignment_id})
+    if self.ta_visible_changed?
```

Comment:
Use `id_changed?` in addition to this condition to also allow for the possibility that a criterion is being created.